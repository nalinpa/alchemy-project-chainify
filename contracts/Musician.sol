// SPDX-License-Identifier: MIT
pragma solidity ^0.8.20;

import "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
import "@openzeppelin/contracts/access/Ownable.sol";

contract Musician is ERC721URIStorage, Ownable {
    uint256 private _nextTokenId = 1;
    mapping(address => bool) public hasMinted;
    uint256 public constant MINT_FEE = 0.01 ether;
    event MusicianMinted(address indexed musician, uint256 tokenId, string tokenURI);

    constructor() ERC721("Musician", "MSC") Ownable(msg.sender) {}

    /// @notice Mint a soulbound NFT for a musician (requires 0.01 ETH)
    /// @param tokenURI Metadata URI for the NFT
    function mint(string memory tokenURI) external payable {
        require(!hasMinted[msg.sender], "Musician: Address already has an NFT");
        require(msg.value >= MINT_FEE, "Musician: Insufficient minting fee");

        uint256 tokenId = _nextTokenId++;
        _safeMint(msg.sender, tokenId);
        _setTokenURI(tokenId, tokenURI);
        hasMinted[msg.sender] = true;

        emit MusicianMinted(msg.sender, tokenId, tokenURI);

        // Refund excess Ether if overpaid
        if (msg.value > MINT_FEE) {
            payable(msg.sender).transfer(msg.value - MINT_FEE);
        }
    }

    /// @notice Override to make tokens soulbound (non-transferable)
    function _update(
        address to,
        uint256 tokenId,
        address auth
    ) internal override returns (address) {
        address from = _ownerOf(tokenId);
        require(from == address(0), "Musician: Token is soulbound");
        return super._update(to, tokenId, auth);
    }

    /// @notice Withdraw contract balance to the owner
    function withdraw() external onlyOwner {
        uint256 balance = address(this).balance;
        require(balance > 0, "Musician: No funds to withdraw");
        payable(owner()).transfer(balance);
    }

    /// @notice Check if an address owns a Musician NFT
    function hasMusicianNFT(address user) external view returns (bool) {
        return hasMinted[user];
    }

    /// @notice Get the last minted token ID
    function getLastMintedTokenId() external view returns (uint256) {
        return _nextTokenId - 1;
    }
}