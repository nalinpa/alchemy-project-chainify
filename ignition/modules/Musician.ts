import { buildModule } from "@nomicfoundation/hardhat-ignition/modules";

export default buildModule("MusicianModule", (m) => {
  const musician = m.contract("Musician");

  return { musician };
});
