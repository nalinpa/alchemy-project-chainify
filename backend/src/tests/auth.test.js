import request from "supertest";
import { app, provider } from "./setup.js"; 
import { ethers } from "ethers";
import jwt from "jsonwebtoken"; 
import User from "../models/user.model.js";
import Musician from "../models/musician.model.js"; 


describe("Authentication API (/api/auth)", () => {
  const userAddress = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
  const userAddressLowerCase = userAddress.toLowerCase();
  const otherUserAddress = "0x70997970C51812dc3A010C7d01b50e0d17dc79C8";
  const loginMessage = "Login to Music NFT Platform";

  let blockchainSnapshotId;
  beforeEach(async () => {
    if (provider && provider.send) {
      blockchainSnapshotId = await provider.send("evm_snapshot", []);
    }
    await User.deleteMany({});
    await Musician.deleteMany({});
  });

  afterEach(async () => {
    if (blockchainSnapshotId && provider && provider.send) {
      await provider.send("evm_revert", [blockchainSnapshotId]);
    }
  });

  describe("POST /api/auth/login", () => {
    it("should login successfully with valid signature and match snapshot", async () => {
      const signer = await provider.getSigner(userAddress);
      const signature = await signer.signMessage(loginMessage);
      const response = await request(app)
        .post("/api/auth/login")
        .send({ address: userAddress, signature: signature })
        .expect(200);
      
      expect(response.body.success).toBe(true);
      expect(response.body.message).toBe("Login successful."); 
      expect(response.body.data.token).toEqual(expect.any(String));
      expect(response.body.data.address.toLowerCase()).toEqual(userAddressLowerCase);
      
      expect(response.body).toMatchSnapshot({
        data: {
          token: expect.any(String),
          pic: expect.any(String),
          userId: expect.any(String), 
          address: expect.any(String) 
        }
      });
      const user = await User.findOne({ address: userAddressLowerCase });
      expect(user).toBeTruthy();
    });

     it("should return 400 for missing address or signature and match error snapshot", async () => {
      const response = await request(app).post("/api/auth/login").send({}).expect(400);
      expect(response.body).toEqual({
        success: false,
        error: { code: "MISSING_CREDENTIALS", message: "Address and signature are required" }
      });
      expect(response.body).toMatchSnapshot();
    });

    it("should return 400 for invalid address and match error snapshot", async () => {
        const signer = await provider.getSigner(userAddress);
        const signature = await signer.signMessage(loginMessage);
        const response = await request(app).post("/api/auth/login").send({ address: "invalid_address", signature }).expect(400);
        expect(response.body).toEqual({
          success: false,
          error: { code: "INVALID_ADDRESS_FORMAT", message: "Invalid wallet address" }
        });
        expect(response.body).toMatchSnapshot();
    });

    it("should return 401 for invalid signature (malformed) and match error snapshot", async () => {
      const response = await request(app).post("/api/auth/login").send({ address: userAddress, signature: "thisisnotavalidsignature" }).expect(401);
      expect(response.body).toEqual({
        success: false,
        error: { 
          code: "SIGNATURE_VERIFICATION_ERROR", 
          message: "Invalid signature format or data.", 
          details: { originalError: expect.any(String) }
        }
      });
      expect(response.body).toMatchSnapshot({error: {details: {originalError: expect.any(String)}}});
    });

     it("should return 401 for signature from wrong address and match error snapshot", async () => {
      const signer = await provider.getSigner(otherUserAddress);
      const signatureFromOther = await signer.signMessage(loginMessage);
      const response = await request(app).post("/api/auth/login").send({ address: userAddress, signature: signatureFromOther }).expect(401);
      expect(response.body).toEqual({
        success: false,
        error: { code: "SIGNATURE_ADDRESS_MISMATCH", message: "Signature does not match the provided address."}
      });
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("POST /api/auth/logout", () => {
    it("should logout successfully and match snapshot", async () => {
      const response = await request(app).post("/api/auth/logout").send({ address: userAddress }).expect(200);
      expect(response.body).toEqual({ success: true, message: "Logged out successfully" });
      expect(response.body).toMatchSnapshot();
    });

     it("should return 400 for missing address on logout and match error snapshot", async () => {
        const response = await request(app).post("/api/auth/logout").send({}).expect(400);
        expect(response.body).toEqual({
          success: false,
          error: { code: "LOGOUT_MISSING_ADDRESS", message: "Address is required for logout" }
        });
        expect(response.body).toMatchSnapshot();
    });
  });
});
