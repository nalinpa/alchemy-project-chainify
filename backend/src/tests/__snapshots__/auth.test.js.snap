// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication API (/api/auth) POST /api/auth/login should login successfully with valid signature and match snapshot 1`] = `
{
  "data": {
    "address": Any<String>,
    "boughtAlbums": [],
    "isMusician": false,
    "pic": Any<String>,
    "token": Any<String>,
    "userId": Any<String>,
  },
  "message": "Login successful.",
  "success": true,
}
`;

exports[`Authentication API (/api/auth) POST /api/auth/login should return 400 for invalid address and match error snapshot 1`] = `
{
  "error": {
    "code": "INVALID_ADDRESS_FORMAT",
    "message": "Invalid wallet address",
  },
  "success": false,
}
`;

exports[`Authentication API (/api/auth) POST /api/auth/login should return 400 for missing address or signature and match error snapshot 1`] = `
{
  "error": {
    "code": "MISSING_CREDENTIALS",
    "message": "Address and signature are required",
  },
  "success": false,
}
`;

exports[`Authentication API (/api/auth) POST /api/auth/login should return 401 for invalid signature (malformed) and match error snapshot 1`] = `
{
  "error": {
    "code": "SIGNATURE_VERIFICATION_ERROR",
    "details": {
      "originalError": Any<String>,
    },
    "message": "Invalid signature format or data.",
  },
  "success": false,
}
`;

exports[`Authentication API (/api/auth) POST /api/auth/login should return 401 for signature from wrong address and match error snapshot 1`] = `
{
  "error": {
    "code": "SIGNATURE_ADDRESS_MISMATCH",
    "message": "Signature does not match the provided address.",
  },
  "success": false,
}
`;

exports[`Authentication API (/api/auth) POST /api/auth/logout should logout successfully and match snapshot 1`] = `
{
  "message": "Logged out successfully",
  "success": true,
}
`;

exports[`Authentication API (/api/auth) POST /api/auth/logout should return 400 for missing address on logout and match error snapshot 1`] = `
{
  "error": {
    "code": "LOGOUT_MISSING_ADDRESS",
    "message": "Address is required for logout",
  },
  "success": false,
}
`;
